Fix for justnimbus integration

kvanzuijlen opened the pull request:
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
The following sensors have been removed from the `justnimbus` integration since they're no longer available through JustNimbus' API:
- Pump starts (pump_starts)
- Pump hours (pump_hours)
- Error code (error_code)
- Total replenished (water_saved)
- Pump flow (pump_flow)
- Drink flow (drink_flow)

The following sensors have been added:
- Pump type (pump_type)

The following sensors have been renamed:
- Total saved (total_saved) -> Total saved (water_saved)
- Total use (totver) -> Total use (water_used)
- Maximum reservoir content (reservoir_content_max) -> (Reservoir capacity) reservoir_capacity

Automations or scripts using any of the removed or renamed entities should be updated. Once updated the old entities can be safely removed.

## Proposed change
Fixes the JustNimbus to work with the new API
https://github.com/kvanzuijlen/justnimbus/compare/0.6.0...0.7.3

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #96948
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28713

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

kvanzuijlen committed:
Fix for justnimbus integration

kvanzuijlen committed:
Merge branch 'dev' into fix-justnimbus

MartinHjelmare commented on homeassistant/const.py:
```diff
@@ -273,6 +273,7 @@ class Platform(StrEnum):
 CONF_WEEKDAY: Final = "weekday"
 CONF_WHILE: Final = "while"
 CONF_WHITELIST: Final = "whitelist"
+CONF_ZIP_CODE: Final = "zip_code"
```
Please move this constant to a const.py module in the justnimbus integration. Two few integrations are using it for it to be made common.

MartinHjelmare reviewed (changes requested):
Tests need to be updated too.

kvanzuijlen committed:
Fixed tests and moved const

kvanzuijlen commented:
Thx for your review. I messed something up with my local tests, sorry for that. Reproduced that they were indeed failing and fixed them now.
Also moved the ZIP code to the JustNimbus const.py file

MartinHjelmare commented on homeassistant/components/justnimbus/config_flow.py:
```diff
@@ -12,13 +12,14 @@
 STEP_USER_DATA_SCHEMA = vol.Schema(
     {
         vol.Required(CONF_CLIENT_ID): cv.string,
+        vol.Required(CONF_ZIP_CODE): cv.string,
```
Existing config entries need to be migrated. Probably start a re-auth flow in `async_setup_entry` and update the config entry data in the re-auth flow in the config flow.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication

MartinFinkenflugel commented:
Hi @kvanzuijlen . Any chance you have some more time to look at this before the PR gets closed?

kvanzuijlen commented:
I'll have a look coming week!

kvanzuijlen committed:
Merge branch 'dev' into fix-justnimbus

kvanzuijlen committed:
Merge branch 'dev' into fix-justnimbus

kvanzuijlen committed:
fix: added reauth flow

kvanzuijlen committed:
fix: fixed reauth config flow

kvanzuijlen committed:
chore: added config_flow reauth test

kvanzuijlen commented on tests/components/justnimbus/test_config_flow.py:
```diff
@@ -103,15 +102,48 @@ async def _set_up_justnimbus(hass: HomeAssistant, flow_id: str) -> None:
     assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_reauth_flow(hass: HomeAssistant) -> None:
```
@MartinHjelmare My Python is a bit rusty :/. I know the test is incorrect but I cannot find out how to correctly test the code. Can you help me by giving me a few pointers?

MartinHjelmare replied:
The test looks ok in general. I added some suggestions for completeness.

MartinHjelmare commented on tests/components/justnimbus/test_config_flow.py:
```diff
@@ -102,17 +102,48 @@ async def _set_up_justnimbus(hass: HomeAssistant, flow_id: str) -> None:
+        return_value=False,
+    ):
+        mock_config = MockConfigEntry(
+            domain=DOMAIN, unique_id=FIXTURE_UNIQUE_ID, data=FIXTURE_USER_INPUT
```
Create the mock config entry with the old data items.

MartinHjelmare commented on tests/components/justnimbus/test_config_flow.py:
```diff
@@ -102,17 +102,48 @@ async def _set_up_justnimbus(hass: HomeAssistant, flow_id: str) -> None:
+                "source": config_entries.SOURCE_REAUTH,
+                "entry_id": mock_config.entry_id,
+            },
+            data=FIXTURE_USER_INPUT,
```
Same here.

MartinHjelmare commented on tests/components/justnimbus/test_config_flow.py:
```diff
@@ -102,17 +102,48 @@ async def _set_up_justnimbus(hass: HomeAssistant, flow_id: str) -> None:
+        await hass.async_block_till_done()
+
+        assert result2["type"] == data_entry_flow.FlowResultType.ABORT
+        assert result2["reason"] == "reauth_successful"
```
Also assert that the mock config entry data was updated with the new data.

MartinHjelmare commented on homeassistant/components/justnimbus/__init__.py:
```diff
@@ -3,14 +3,18 @@
-    coordinator = JustNimbusCoordinator(hass=hass, entry=entry)
+    try:
+        coordinator = JustNimbusCoordinator(hass=hass, entry=entry)
+    except KeyError as err:
```
I'd explicitly check the config entry data for the new key so we don't need to catch a general exception like KeyError here.

kvanzuijlen replied:
Yeah, you're right, this is a bit nicer than catching a general exception. Thx for the great feedback and suggestions! I have updated the PR!

MartinHjelmare commented on homeassistant/components/justnimbus/sensor.py:
```diff
@@ -104,57 +69,46 @@ class JustNimbusEntityDescription(
-    JustNimbusEntityDescription(
-        key="totver",
-        translation_key="total_use",
+        key="water_used",
```
We need to add a breaking change paragraph for removed and added sensors. Explain what the user needs to do to cope with the changes. The user can delete the old entities after updating any automations or scripts that use them.

MartinHjelmare commented on homeassistant/components/justnimbus/__init__.py:
```diff
@@ -3,14 +3,18 @@
+    try:
+        coordinator = JustNimbusCoordinator(hass=hass, entry=entry)
+    except KeyError as err:
+        raise ConfigEntryAuthFailed(err) from err
```
Test this case by adding a mock config entry with old data and set it up. Then check that there's a reauth flow in progress. Add this test to `test_init.py`.

kvanzuijlen committed:
chore: Processed PR feedback

emontnemery commented:
The breaking change section should explain in one to two sentences why the sensors had to be removed

MartinHjelmare commented:
@emontnemery are you happy with the breaking change section now?

gjohansson-ST committed:
Merge branch 'dev' into fix-justnimbus
